### ISSUE TITLE
Add the ability to use out on Crystal methods and procs

### ISSUE THREAD

**Issue by mjblack** [crystal-lang/crystal#13799]
#### _Is your feature request related to a problem? Please describe clearly and concisely what is it._

Like Lib funs it would be great if we could use `out` keyword in parameters to methods and procs outside of Libs. Interfacing with COM classes in Windows requires that the VTable struct to contain fields that are Proc type and many of the methods have `out` direction on the params, the returned value being sucess or error code. 

https://learn.microsoft.com/en-us/windows/win32/api/rometadataapi/nf-rometadataapi-imetadatatables-getcodedtokeninfo 

#### _Describe the feature you would like, optionally illustrated by examples, and how it will solve the above problem._

The feature would allow for methods and procs to be able to accept `out` for params. 

#### _Describe considered alternative solutions, and the reasons why you have not proposed them as a solution here._

Right now the only solution is to initialize pointers ahead of time and pass them as parameters.

#### _Does it break backward compatibility, if yes then what's the migration path?_

This feature should not affect backward compatibility.

**Comment by oprypin** [comment]
> Right now the only solution is to initialize pointers ahead of time and pass them as parameters.

No? You just return multiple values in a tuple or another structure and that's it

> many of the methods have `out` direction on the params, the returned value being sucess or error code.

If you're making convenient wrappers for Crystal users, then dealing with separate numeric codes doesn't count as "convenient", you should change them into expressions or at least as a union type `Result | ErrorCode`.
If you're not making convenient wrappers, then just leave the `fun` as is and it can be used directly.

This will not be done, it greatly reduces obviousness of the language because you can't be sure anymore how a parameter will behave when you see one being passed.

**Comment by oprypin** [comment]
Oh sorry, the last part that I wrote is wrong - there wouldn't be a surprise for the caller side if the `out` keyword is present there of course

**Comment by mjblack** [comment]
> > Right now the only solution is to initialize pointers ahead of time and pass them as parameters.
> 
> 
> 
> No? You just return multiple values in a tuple or another structure and that's it
>
> 

I would but the function pointer is created by the C/C++ library and it is not returning a tuple.

> If you're not making convenient wrappers, then just leave the `fun` as is and it can be used directly.

It's not a fun being called but function pointers.

**Comment by oprypin** [comment]
I see that you're dealing with function pointers and not necessarily even `fun`s. In that case indeed you have to create the variable in advance and can't use `out`

> it is not returning a tuple.

Yes but in your wrapper you can receive data this way, transform it and return a tuple

**Comment by mjblack** [comment]
> > it is not returning a tuple.
> 
> Yes but in your wrapper you can receive data this way, transform it and return a tuple

Right, but creating those wrappers can be very time consuming to create with large libraries like Win32 API.

**Comment by oprypin** [comment]
That should be automated then :/
